### ISSUE TITLE
Helm 3 support

### ISSUE THREAD

**Issue by donbeave** [werf/werf#1606]
Hello, guys!

Amazing project! Looks like exactly what I’m looking for.

Just some probably weird question for me, if Helm 3 is supported? I know it’s still alpha, but it still works well for us even in this stage, will be happy to use werf with third version of Helm. If not supported yet, pls let me know your roadmap to upgrade to next version of Helm.

Best regards,
Alexey Zhokhov.

**Comment by distorhead** [comment]
Hi! Thanks!

Werf now have built-in helm 2 client and tiller. It works the same way as helm 2 (same release storage, fully compatible, etc.). But the only difference is that tiller process and werf process are the same single process. So to use helm 3 we must compile werf agains new helm 3 codebase.

Werf is not compiled with helm 3 for now. The main reasons are: there should be clear and durable way to migrate from helm 2 installations to helm 3 (https://github.com/helm/helm/issues/5892); it is alpha.

We will certainly update to helm 3 some day.

Public roadmap will also be available at the https://werf.io soon.

**Comment by donbeave** [comment]
@distorhead https://github.com/helm/helm/releases/tag/v3.0.0 just released!

**Comment by StupidScience** [comment]
Any plans/ETA here?

**Comment by javad-hajiani** [comment]
👍

**Comment by distorhead** [comment]
Some notes for future implementation:
 - werf should not auto-purge first failed release, remove this dangerous feature when ported to helm 3;
 - update masterminds/sprig library;
 - CTRL-C + wait + atomic release status updates;
 - 3 way merge should solve a problem when a kubernetes cluster has been updated, but there are resources deployed which have outdated api versions definitions in the release.

**Comment by distorhead** [comment]
> Any plans/ETA here?

It is not a priority task for now, because it will add nothing new and valuable to the werf. Werf do have all main features that Helm 3 have: three way merge and tillerless architecture.

But we should switch to Helm 3 in v1.2 version.

**Comment by Menzorg** [comment]
Good day. Thanks, could you say, when 1.2 is planning?

**Comment by distorhead** [comment]
**Update.** Happy to announce, that we have "migration to Helm 3" task in the nearest plans of big tasks. I think we will implement usable Helm 3 support in 1 month approximately.

**Comment by distorhead** [comment]
As for now, there is very early alpha version: v1.1.22-alpha4 — which available in the alpha channel. This version embeds all helm-v3 commands into `werf helm-v3` subcommand and adds support for kubedog-based waiting for resources. For example:

```
werf helm-v3 upgrade --namespace myproj myproj . --install --create-namespace --wait
```

All werf deploy-process related annotations are supported as well: https://werf.io/documentation/reference/deploy_process/deploy_into_kubernetes.html#resource-tracking-configuration.

This can already be useful for someone. But we continue to work on full helm-3 support (secrets, support replicas autoscaler for 3-way-merge managed fields, resources adoption, automatically added annotations and labels to all deployed resources, etc.).

**Comment by distorhead** [comment]
Status update.

As for now Helm 3 is experimentally avaiable in the current v1.1 version when `WERF_DEPLOY_HELM_V3=1` variable is set — werf will use helm-v3 for werf deploy/dismiss commands in this case.

Also we are working on werf v1.2 which is in alpha stage now. It use helm-v3 by default. There is unsolved question with automigration from existing helm-v2 installation to helm-v3 now.

Usable v1.2 should be available in october.

**Comment by distorhead** [comment]
Here we go. Alpha and beta channel of v1.2 now contain usable version of werf with Helm 3.4.0.

 - `werf converge` command will build images defined in the werf.yaml and deploy chart using helm 3.
 - v1.2 uses helm 3 only and does not support helm 2.
 - checkout new documentation articles: https://werf.io/introduction.html, https://werf.io/documentation/quickstart.html and https://werf.io/documentation/using_with_ci_cd_systems.html.
 - `werf helm *` commands are available to work directly with helm-chart by lowlevel commands (such as install/upgrade and other), these commands use helm 3 only.


As for automigration and further steps:
 1. Current beta version of werf can be used for new projects which does not need a migration from Helm 2.
 2. We are working on this feature https://github.com/werf/werf/issues/2712 to publish charts bound with correct images by semver into docker registry. This feature will already be available in the v1.2 when v1.2 becomes stable.
 3. Next step after releasing first stable v1.2 is to add automigration existing releases from helm 2 to helm 3: https://github.com/werf/werf/issues/2866.
